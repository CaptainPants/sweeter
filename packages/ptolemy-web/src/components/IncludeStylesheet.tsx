/* @jsxImportSource .. */

import {
    $derived,
    $val,
    type Component,
    type MightBeSignal,
    type Signal,
} from '@serpentis/ptolemy-core';
import { arrayExcept } from '@serpentis/ptolemy-utilities';

import { getWebRuntime } from '../runtime/getWebRuntime.js';
import {
    type AbstractGlobalCssStylesheet,
    type StylesheetInclude,
} from '../styles/index.js';

export interface IncludeStylesheetProps {
    stylesheet: StylesheetInclude;
};

export const IncludeStylesheet: Component<IncludeStylesheetProps> = (
    { stylesheet },
    init,
) => {
    const webRuntime = getWebRuntime();

    let previousTime: AbstractGlobalCssStylesheet[] | undefined;

    const flattened = flatten(stylesheet);

    // Remember that this is called on mount as well as when the subscribed signals change.
    init.trackSignals([flattened], ([thisTime]) => {
        if (previousTime) {
            const added = arrayExcept(thisTime, previousTime);
            const removed = arrayExcept(previousTime, thisTime);

            for (const addedItem of added) {
                webRuntime.addStylesheet(addedItem);
            }
            for (const removedItem of removed) {
                webRuntime.removeStylesheet(removedItem);
            }
        } else {
            for (const addedItem of thisTime) {
                webRuntime.addStylesheet(addedItem);
            }
        }

        previousTime = thisTime;
    });

    init.onUnMount(() => {
        if (previousTime) {
            for (const removedItem of previousTime) {
                webRuntime.removeStylesheet(removedItem);
            }
        }
    });

    return <></>;
};

function flatten(
    value: MightBeSignal<StylesheetInclude>,
): Signal<AbstractGlobalCssStylesheet[]> {
    return $derived(() => {
        value = $val(value);

        if (Array.isArray(value)) {
            return value;
        } else {
            return [value];
        }
    });
}
