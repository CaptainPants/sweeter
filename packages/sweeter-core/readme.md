# Sweeter Core
This package contains the share core of Sweeter. 

This includes:
- Signals
- Base Components
- Base Hooks
- Context