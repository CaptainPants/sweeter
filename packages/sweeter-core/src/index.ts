import './jsx-globals.js';
import './SweeterExtensionPoints.js';

export * from './addExplicitStrongReference.js';
export * from './components/index.js';
export * from './context/index.js';
export * from './executionContext/index.js';
export * from './lazy/index.js';
export * from './middleware/index.js';
export * from './runtime/index.js';
export * from './signals/index.js';
export * from './types.js';
export * from './test/index.js';
export * from './utility/index.js';