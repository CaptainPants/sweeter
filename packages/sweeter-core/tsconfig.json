{
    "extends": "../../common-config/tsconfig.json",
    "include": [
        "src/**/*.ts",
        "src/**/*.tsx"
    ],
    "compilerOptions": {
        "rootDir": "src",
        "lib": [
            "ES2022", // No DOM here
        ],
        "noEmit": true,
        "paths":{
        }
    }
}
