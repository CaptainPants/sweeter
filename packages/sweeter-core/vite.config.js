import sweeterPlugin, { watchDependenciesPlugin } from '@captainpants/rollup-plugin-sweeter';
import { resolve } from 'path';
import nodeExternals from 'rollup-plugin-node-externals';
import { defineConfig } from 'vite';
import circularDependency from 'vite-plugin-circular-dependency';
import dts from 'vite-plugin-dts';

export default defineConfig({
    build: {
        lib: {
            // Could also be a dictionary or array of multiple entry points
            // eslint-disable-next-line no-undef
            entry: resolve(__dirname, './src/index.ts'),
            fileName: 'index',
            name: 'index',
            formats: ['es'],
        },
        outDir: 'dist',
        minify: false,
        rollupOptions: {
        },
        target: "ESNext",
    },
    test: {
        poolOptions: {
            forks: {
                execArgv: ["--expose-gc"]
            }
        },
    },
    plugins: [
        dts({ clearPureImport: false }), 
        nodeExternals(), 
        circularDependency({ circleImportThrowErr: true }), 
        sweeterPlugin({ projectName: '@captainpants/sweeter-core', roots: [__dirname] }),
        watchDependenciesPlugin({ dependencies: [{ namePattern: '@captainpants/*', filesPattern: 'dist/built.txt', buildCompleteNoticePath: 'dist/built.txt' }] })
    ]
});
