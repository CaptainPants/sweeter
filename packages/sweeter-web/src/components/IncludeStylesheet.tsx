/* @jsxImportSource .. */

import type { Component } from '@captainpants/sweeter-core';
import type { GlobalCssStylesheet } from '../styles/index.js';
import { getWebRuntime } from '../runtime/getWebRuntime.js';

export interface IncludeStylesheetProps {
    stylesheet: GlobalCssStylesheet;
}

export const IncludeStylesheet: Component<IncludeStylesheetProps> = (
    { stylesheet },
    init,
) => {
    const runtime = getWebRuntime();

    init.subscribeToChanges(
        [stylesheet],
        ([stylesheet]) => {
            if (stylesheet) {
                const remove = runtime.addStylesheet(stylesheet);

                return remove;
            }

            return undefined;
        },
        true,
    );

    return <></>;
};
