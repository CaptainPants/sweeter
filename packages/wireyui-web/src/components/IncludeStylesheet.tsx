/* @jsxImportSource .. */

import type { Component } from '@captainpants/wireyui-core';
import type { AbstractGlobalCssStylesheet } from '../styles/index.js';
import type { DocumentStylesheetHandle } from '../runtime/WebRuntimeContext.js';
import { WebRuntimeContext } from '../runtime/WebRuntimeContext.js';

export interface IncludeStylesheetProps {
    stylesheet: AbstractGlobalCssStylesheet;
}

export const IncludeStylesheet: Component<IncludeStylesheetProps> = (
    { stylesheet },
    init,
) => {
    const runtime = init.getContext(WebRuntimeContext);

    let handle: DocumentStylesheetHandle | undefined;

    init.subscribeToChanges(
        [stylesheet],
        ([stylesheet]) => {
            if (handle) {
                if (stylesheet) {
                    handle.update(stylesheet);
                } else {
                    handle.remove();
                }
            } else {
                if (stylesheet) {
                    handle = runtime.addStylesheet(stylesheet);
                }
            }
            return;
        },
        true,
    );

    init.onUnMount(() => {
        handle?.remove();
    });

    return <></>;
};
