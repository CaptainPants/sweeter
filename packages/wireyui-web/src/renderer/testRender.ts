import { WebRenderer } from './WebRenderer.js';

export interface TestRenderResult {
    readonly nodes: NodeListOf<ChildNode>;
    readonly getHTML: () => string;
    readonly dispose: () => void;
}

/**
 * Convenience function for rendering isolated JSX chunks for unit tests
 * @param callback
 * @returns
 */
export function testRender(callback: () => JSX.Element): TestRenderResult {
    const renderer = new WebRenderer();

    const root = document.createElement('div');

    const dispose = renderer.start(root, callback);

    return {
        nodes: root.childNodes,
        getHTML: () => root.innerHTML,
        dispose,
    };
}
