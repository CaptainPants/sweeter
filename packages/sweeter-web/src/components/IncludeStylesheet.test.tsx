/* @jsxImportSource .. */

import { type Component } from '@captainpants/sweeter-core';

import { GlobalCssClass } from '../styles/index.js';
import { testRender } from '../test/testRender.js';

import { IncludeStylesheet } from './IncludeStylesheet.js';

const example = new GlobalCssClass({
    className: 'Something',
    content: `
        body { background-color: red; }
    `,
});

it('Global stylesheet added', () => {
    const res = testRender(() => {
        return <IncludeStylesheet stylesheet={example} />;
    });

    expect(document.head.innerHTML).toMatchSnapshot();

    res.dispose();

    // Make sure it was removed again
    expect(document.head.innerHTML).toStrictEqual('');
});

const component1Stylesheet = new GlobalCssClass({
    className: 'Component1',
    content: `
        color: blue;
    `,
});

const Component1: Component = (_props, _init) => {
    return (
        <>
            <IncludeStylesheet stylesheet={component1Stylesheet} />
            <div class={component1Stylesheet}>This is an example</div>
        </>
    );
};

it('Global style for component', () => {
    const res = testRender(() => {
        return <Component1 />;
    });

    expect(document.head.innerHTML).toMatchSnapshot();

    expect(res.getHTML()).toMatchSnapshot();

    res.dispose();
});

const component2Stylesheet = new GlobalCssClass({
    className: 'Component2',
    content: `
        color: blue;

        .inner {
            background-color: green;
        }
    `,
});

const Component2: Component = (_props, _init) => {
    return (
        <>
            <IncludeStylesheet stylesheet={component2Stylesheet} />
            <div class={component2Stylesheet}>
                This is an example
                <div class="inner" />
            </div>
        </>
    );
};

it('Global style for component with nested', () => {
    const res = testRender(() => {
        return <Component2 />;
    });

    expect(document.head.innerHTML).toMatchSnapshot();

    expect(res.getHTML()).toMatchSnapshot();

    res.dispose();
});
