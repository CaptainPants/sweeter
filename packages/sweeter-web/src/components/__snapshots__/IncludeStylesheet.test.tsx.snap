// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global style for component 1`] = `
"<style data-id="Component1">
._glbl0_Component1 {
    color: blue;
}
</style>"
`;

exports[`Global style for component 2`] = `"<!--Hooks(IncludeStylesheet): Mount, UnMount--><div class="_glbl0_Component1">This is an example</div>"`;

exports[`Global style for component with nested 1`] = `
"<style data-id="Component2">
._glbl0_Component2 {
    color: blue;
}
._glbl0_Component2 .inner {
    background-color: green;
}
</style>"
`;

exports[`Global style for component with nested 2`] = `"<!--Hooks(IncludeStylesheet): Mount, UnMount--><div class="_glbl0_Component2">This is an example<div class="inner"></div></div>"`;

exports[`Global stylesheet added 1`] = `
"<style data-id="Something">
._glbl0_Something body {
    background-color: red;
}
</style>"
`;
